### PR TITLE
feat: add microsoft calendar/teams integration

### LINKED ISSUE #24: Add Meeting Scheduling component for Microsoft Teams

### PR DESCRIPTION
**Pull request by askmrsinh**
closes https://github.com/wxsd-sales/roomos-device-widgets/issues/24

### CODE REVIEW

File: src/routes/api/client-credentials/microsoft/token.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+  // validate request header
+  const obj = humps.camelizeKeys(Object.fromEntries(requestEvent.request.headers.entries()));
+  const header = plainToInstance(RequestHeaderDTO, obj, classTransformOptions);
+  const headerValidationErrors = validateSync(header, classValidationOptions);
+  if (headerValidationErrors.length > 0) {
+    return { status: 401 };
+  }
+
+  const orm = await MikroORM.init({ ...config, ...{ entities: [Activation] } });
```
Comment (akoushke) [inline src/routes/api/client-credentials/microsoft/token.ts:94]: Can't you just access the entity attached to the request Event local property instead of initiating another Driver?

File: src/routes/api/client-credentials/microsoft/token.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    .getMyOwnDetails()
+    .then((r) => r.json())
+    .then((r) => r.emails[0]);
+
+  const account = em.findOneOrFail(Activation, query.activationId, {
+    fields: ['microsoftTenantId', 'microsoftClientId', 'microsoftPrivateKey', 'microsoftClientCertificate']
+  });
+
+  const formJwt = (audience: string, subject: string, issuer: string, pkcs8: string, x509: string, expiry = '1h') => {
+    const alg = 'RS256';
```
Comment (akoushke) [inline src/routes/api/client-credentials/microsoft/token.ts:107]: can you make this an env variable?

File: src/routes/api/client-credentials/microsoft/token.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    const data = x509
+      .replaceAll('\n', '')
+      ?.match(/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/gm)
```
Comment (akoushke) [inline src/routes/api/client-credentials/microsoft/token.ts:110]: global constant perhaps?

File: src/routes/api/client-credentials/microsoft/token.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        .sign(r2)
+    );
+  };
+
+  const addExpiresAt = (obj: JSONObject, date: number = Date.now()) => ({
+    ...obj,
+    ...{ expiresAt: date + (obj.expiresIn as number) * 1000 }
+  });
+
+  return Promise.all([subject, account])
```
Comment (akoushke) [inline src/routes/api/client-credentials/microsoft/token.ts:133]: in couple of sentences can you explain what this promise resolve? nested promises are kinda hard to follow IMO.

File: src/routes/demos/activate/.Step5MicrosoftIntegration.svelte [hunk 5]
```diff
@@ -0,0 +1,129 @@
+<script lang="ts">
+  export let tenantId;
+  export let clientId;
+  export let privateKey;
+  export let clientCertificate;
+
+  const logo = 'https://cdn.jsdelivr.net/npm/simple-icons@v7/icons/microsoft.svg';
```
Comment (akoushke) [inline src/routes/demos/activate/.Step5MicrosoftIntegration.svelte:7]: $lib/constant maybe? if it's only being used here, your call.

File: src/routes/demos/activate/.Step5MicrosoftIntegration.svelte [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  const privateKeyPlaceholder = '-----BEGIN PRIVATE KEY-----' + '\n\n\n\n\n\n' + '-----END PRIVATE KEY-----';
```
Comment (akoushke) [inline src/routes/demos/activate/.Step5MicrosoftIntegration.svelte:8]: I think this will be used in other places too. maybe a global const

File: src/routes/demos/activate/.Step5MicrosoftIntegration.svelte [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        placeholder={privateKeyPlaceholder}
+        required={isRequired}
+        bind:value={privateKey}
+      />
+      <span class="icon is-right">
+        <i class="mdi mdi-certificate" />
+      </span>
+    </div>
+    <div class="help">
+      <p>This is the private key for your app registration's certificate</p>
```
Comment (akoushke) [inline src/routes/demos/activate/.Step5MicrosoftIntegration.svelte:103]: you can you escape out the `'`. it messing up the syntax coloring in github.

File: src/routes/demos/activate/index.ts [hunk 8]
```diff
@@ -31,22 +31,48 @@ export const POST = async (requestEvent: RequestEvent) => {
     public readonly demoUuid!: string;
 
     @Expose()
-    @ValidateIf(({ obj }) => obj?.googlePrivateKey != null || obj?.googleClientCertificate != null)
+    @ValidateIf(({ obj }) => obj?.googlePrivateKey ?? obj?.googleClientCertificate)
```
Comment (akoushke) [inline src/routes/demos/activate/index.ts:34]: what do two question marks mean?

**Review by akoushke — CHANGES_REQUESTED** [review]
Please see the comments.

File: src/routes/api/client-credentials/microsoft/token.ts (hunk 1 above)
Comment (askmrsinh) [reply src/routes/api/client-credentials/microsoft/token.ts:94]: Good point. But `/api` routes are not attached to a session and also don't have access to `event.locals.db`. The API routes that are dependent on the DB are very limited (3 at present, including this one). Even with that, those 3 need access to at most 1 or 2 tables (entities). Hence initialization is done on case by case basis.

File: src/routes/api/client-credentials/microsoft/token.ts (hunk 2 above)
Comment (askmrsinh) [reply src/routes/api/client-credentials/microsoft/token.ts:107]: Will make this a constant as part of PR https://github.com/wxsd-sales/roomos-device-widgets/pull/36.

File: src/routes/api/client-credentials/microsoft/token.ts (hunk 3 above)
Comment (askmrsinh) [reply src/routes/api/client-credentials/microsoft/token.ts:110]: Will make this a constant as part of PR https://github.com/wxsd-sales/roomos-device-widgets/pull/36.

File: src/routes/api/client-credentials/microsoft/token.ts (hunk 4 above)
Comment (askmrsinh) [reply src/routes/api/client-credentials/microsoft/token.ts:133]: Essentially generates a JWT. The flow is described here: https://stackoverflow.com/a/75168441.

File: src/routes/demos/activate/.Step5MicrosoftIntegration.svelte (hunk 5 above)
Comment (askmrsinh) [reply src/routes/demos/activate/.Step5MicrosoftIntegration.svelte:7]: Currently, this is the only usage. I plan to add https://www.npmjs.com/package/simple-icons as a dependency.

File: src/routes/demos/activate/.Step5MicrosoftIntegration.svelte (hunk 6 above)
Comment (askmrsinh) [reply src/routes/demos/activate/.Step5MicrosoftIntegration.svelte:8]: It's being used at one other place. Will make this a constant as part of PR https://github.com/wxsd-sales/roomos-device-widgets/pull/36.

File: src/routes/demos/activate/.Step5MicrosoftIntegration.svelte (hunk 7 above)
Comment (askmrsinh) [reply src/routes/demos/activate/.Step5MicrosoftIntegration.svelte:103]: Thanks. There are other instances of this across the whole code base. Will refactor as part of PR https://github.com/wxsd-sales/roomos-device-widgets/pull/36.

File: src/routes/demos/activate/index.ts (hunk 8 above)
Comment (askmrsinh) [reply src/routes/demos/activate/index.ts:34]: It's the relatively new [nullish coalescing (??) operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing).

**Review by akoushke — APPROVED** [review]
Some minor changes which will be added to upcoming PR
